### PR TITLE
Reduce help handler wall of text

### PR DESCRIPTION
**Pull request by t33chong**
This change allows help topics to be filtered by handler and an optional
substring.

### CODE REVIEW

**Comment by t33chong** [comment]
What the proposed change looks like using the shell adapter:

```
Lita > !help
Type '!help HANDLER' to list terms or commands for that handler. You may optionally specify a substring on which to filter after the handler name. The following handlers are installed:
ascii_art
authorization
broadcast
help
info
log
room
users
Lita > !help help
!help - Lists help information for terms and command the robot will respond to.
!help HANDLER - Lists available commands for HANDLER.
!help HANDLER [COMMAND] - Lists help information for terms or commands in HANDLER that match COMMAND.
Lita > !help help match
!help HANDLER [COMMAND] - Lists help information for terms or commands in HANDLER that match COMMAND.
```

**Comment by MrJoy** [comment]
+1

File: lib/lita/handlers/help.rb [hunk 1]
```diff
@@ -28,39 +37,49 @@ def authorized?(user, required_groups)
...
 
-      # Creates an array of help info for all registered routes.
-      def build_help(response)
+      # Creates an alphabetically-sorted array containing the names of all
+      # installed handlers.
+      def list_handlers
         robot.handlers.map do |handler|
-          next unless handler.respond_to?(:routes)
+          handler.namespace if handler.respond_to?(:routes)
+        end.flatten.compact.sort
```
Comment (jimmycuadra) [inline lib/lita/handlers/help.rb]: I believe it's more efficient to use `flat_map` instead of `map {}.flatten`.

File: lib/lita/handlers/help.rb [hunk 2]
```diff
@@ -28,39 +37,49 @@ def authorized?(user, required_groups)
...
+      end
 
+      # Creates an array of help info for a specified handler. Optionally
+      # filters commands matching a given substring.
+      def list_commands(handler_name, substring, user)
+        handlers = robot.handlers.select { |handler| handler.namespace == handler_name.strip.downcase }
+        return ["No matching handlers found for '#{handler_name}'"] if handlers.empty?
```
Comment (jimmycuadra) [inline lib/lita/handlers/help.rb]: This response text should go through i18n.

**Comment by jimmycuadra** [comment]
Great work! Left a couple of inline comments. Other than that, looks good.

**Comment by t33chong** [comment]
@jimmycuadra Thanks! Just addressed your comments.

**Pull request merged**